scheduler/utilization_based_host_allocator.go

```diff
@@ -221,27 +159,66 @@ func evalHostUtilization(ctx context.Context, d distro.Distro, taskQueue []model
 		"new_hosts_needed":             numNewHosts,
 		"num_existing_hosts":           len(existingHosts),
 		"num_free_hosts_approx":        numFreeHosts,
-		"queue_length":                 len(taskQueue),
-		"queue_tasks":                  queueTasks,
-		"long_tasks":                   hostsForLongTasks,
-		"scheduled_tasks_runtime":      int64(scheduledTasksDuration),
-		"scheduled_tasks_runtime_span": scheduledTasksDuration.String(),
+		"queue_length":                 taskGroupInfo.Count,
+		"long_tasks":                   numLongTasks,
+		"scheduled_tasks_runtime":      int64(scheduledDuration),
+		"scheduled_tasks_runtime_span": scheduledDuration.String(),
 		"op_dur_free_host_secs":        freeHostDur.Seconds(),
 		"op_dur_total_secs":            time.Since(evalStartAt).Seconds(),
 	})
 
 	return numNewHosts, nil
 }
 
-// calcScheduledTasksDuration returns the total estimated duration of all
-// tasks scheduled to be run in a given task queue
-func calcScheduledTasksDuration(queue []model.TaskQueueItem) time.Duration {
-	var scheduledTasksDuration time.Duration
+// groupByTaskGroup takes a list of hosts and tasks and returns them grouped by task group
+func groupByTaskGroup(runningHosts []host.Host, distroQueueInfo DistroQueueInfo) map[string]TaskGroupData {
+	taskGroupDatas := map[string]TaskGroupData{}
+	for _, h := range runningHosts {
+		name := ""
+		if h.RunningTask != "" && h.RunningTaskGroup != "" {
+			name = makeTaskGroupString(h.RunningTaskGroup, h.RunningTaskBuildVariant, h.RunningTaskProject, h.RunningTaskVersion)
+		}
+		if data, exists := taskGroupDatas[name]; exists {
+			data.Hosts = append(data.Hosts, h)
+			taskGroupDatas[name] = data
+		} else {
+			taskGroupDatas[name] = TaskGroupData{
+				Hosts: []host.Host{h},
+				Info:  TaskGroupInfo{},
+			}
+		}
+	}
+	taskGroupsInfosMap := distroQueueInfo.taskGroupInfosMap
+	for name, info := range taskGroupsInfosMap {
+		if data, exists := taskGroupDatas[name]; exists {
+			data.Info = info
+			taskGroupDatas[name] = data
+		} else {
+			taskGroupDatas[name] = TaskGroupData{
+				Hosts: []host.Host{},
+				Info:  info,
+			}
+		}
+	}
 
-	for _, taskQueueItem := range queue {
-		scheduledTasksDuration += taskQueueItem.ExpectedDuration
+	// Any task group can use a host not running a task group, so add them to each list.
+	// This does mean that we can plan more than 1 task for a given host from 2 different
+	// task groups, but that should be in the realm of "this is an estimate"
+	if taskGroupData, ok := taskGroupDatas[""]; ok {
+		for name, data := range taskGroupDatas {
+			if name == "" {
+				continue
+			}
+			data.Hosts = append(data.Hosts, taskGroupData.Hosts...)
+			taskGroupDatas[name] = data
+		}
 	}
-	return scheduledTasksDuration
+
+	return taskGroupDatas
+}
+
+func makeTaskGroupString(group, bv, project, version string) string {
```

Comment:
feel like this function should probably be a method on the model?

Reply:
Added separate functions to *task.Task* and *host.Host*